src/ast/passes/semantic_analyser.cpp

```diff
@@ -2038,6 +2034,62 @@ void SemanticAnalyser::visit(For &f)
    *       $kv = ((uint64)key, (uint64)value);
    *       [LOOP BODY]
    *     }
+   *
+   *
+   * To allow variables to be shared between the loop callback and the main
+   * program, some extra steps are taken:
+   *
+   * 1. Determine which variables need to be shared with the loop callback
+   * 2. Pack pointers to them into a context struct
+   * 3. Pass pointer to the context struct to the callback function
+   * 4. In the callback, override the shared variables so that they read and
+   *    write through the context pointers instead of directly from their
+   *    original addresses
+   *
+   * Example transformation with context:
+   *
+   * Before:
+   *     PROBE {
+   *       $str = "hello";
+   *       $not_shared = 2;
+   *       $len = 0;
+   *       @map[11, 12] = "c";
+   *       for ($kv : @map) {
+   *         print($str);
+   *         $len++;
+   *       }
+   *       print($len);
+   *       print($not_shared);
+   *     }
+   *
+   * After:
+   *     struct ctx_t {
```

Comment:
I haven't read the code thoroughly enough yet but is there a possible collision here if there are two (or more) probes that have scratch variables with the same names